2024/incoming/a-time-tested-powerhouse-for-processing-xml.pod

```diff
@@ -0,0 +1,171 @@
+the name of a person we must access name attribute. L<XML::Twig::Elt> has various functions on
+attributes. The C<att> method will do the job. To note that, because the XML we have is automatically
+generated by another system, so we haven't checked if the element has the attribute. To check if an
+element has a specific attribute, you can call the C<att_exist> it will return true if the attribute
+exists for the element, false otherwise.
+
+  #!perl
+
+  sub check_guest {
+    # previous lines above
+
+    my @guests = $twig->root->children('guest');
+    # $guests[0]->att('name') will return Michael
+  }
+
+
+If we put all the information we’ve discussed together, the function would look like this.
+
+  #!perl
+
+  sub check_guest {
+    my $id = shift;
+    my $password = shift; 
+
+    my $twig = XML::Twig->new();
+    $twig->parsefile( "/hotelpms/guests.xml" ) || return 0; # handle error
+    my @guests = $twig->root->children('guest');
+    
+    foreach my $guest (@guests) {
+      return 1 if (($guest -> att('id') eq $id) && ($guest -> att('wifiKey') eq $password));
+    }
+
+    return 0;
+  }
+
+=head2 Parsing XML Response
+
+Suppose that there are servers somewhere in this universe. We don't know exactly when the answer
+will be returned it changes depending on the data of our request, but we know the exact XML response
+format. It is said that this server provides answers about the Earth's past based on what it 
+observes. So, demanding information related to an much earlier date will take much longer to travel
+due to the speed of light and it is open anyone to send request. A researcher wants to find out 
+why the stones at Gobeklitepe were buried. Researcher knows that when the answer reach researcher
+will not be alive. All conditions have been maintained for the message to return, and a system was 
+designed to await all requests by some world organization. The language selected was Perl.
+
+Here is expected response format:
+
+  #!vim xml
+
+  HTTP/1.1 200 OK
+  Content-Type: text/xml; charset=utf-8
+  Content-Length: length
+
+  <?xml version="1.0" encoding="utf-8"?>
+  <soap:Envelope xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:xsd="http://www.w3.org/2001/XMLSchema" xmlns:soap="http://schemas.xmlsoap.org/soap/envelope/">
+    <soap:Body>
+      <observeResult xmlns="http://universe.com/">
+        <query>string</query>
+        <observation>
+          <year></year>
+          <nation>string</nation>
+          <purpose>string</purpose>
+          <details>string</details>
+        </observation>
+      </observeResult>
+    </soap:Body>
+  </soap:Envelope>
+
+
+The C<first_child> function in the XML::Twig module is used to retrieve the first child element of the
+current element in the XML document. L<XML::Twig::Elt> class has C<text> method that extracts the 
+content from inside tag.
+
+  #!perl
+
+  my $ua = LWP::UserAgent->new();
```

Comment:
I think we'll need to add the `use` statements for `HTTP::Request` and `LWP::UserAgent` here.

Reply:
HTTP::Request is pulled in by LWP::UserAgent so that one doesn't need an explicit `use`. I agree about the other two, though.

Reply:
Yes, it is working, I tested it with pseudo response. 

> HTTP::Request is pulled in by LWP::UserAgent so that one doesn't need an explicit `use`. I agree about the other two, though.

I added `HTTP::Request`, `XML::Twig` and `LWP::UserAgent`. When I don't import  `LWP::UserAgent` it gives me an error so I added explicitly.

Reply:
HTTP::Request is the one you can omit. Doesn't hurt keeping it, though.

Reply:
You are right 👍  I removed it now.

Reply:
> HTTP::Request is the one you can omit. Doesn't hurt keeping it, though.

My preference would be to keep it, as I think it makes the code easier to understand. As the maintainer of both `HTTP::Request` and `LWP::UserAgent` it had not even occurred to me that skipping an import would be fine. I'd think others who may not be as close to the code may end up asking themselves where the import came from and that's likely a distraction from the point that the code is trying to make.